### PR TITLE
IDV test for DvTypes

### PR DESCRIPTION
**Pull request by codemzs**
This test ensures IDV file generated using DvTypes is parsed by the new type system that does not contain DvTypes.

### CODE REVIEW

**Comment by TomFinley** [comment]
Great, thanks @codemzs. Parquet?

**Comment by codemzs** [comment]
One step at a time, Sir. I want to nail down IDV before I go to Parquet. 

---
In reply to: [416348763](https://github.com/dotnet/machinelearning/pull/750#issuecomment-416348763) [](ancestors = 416348763)

**Comment by TomFinley** [comment]
In this PR, or are you phrasing an entire separate one for each single test?

---
In reply to: [416349177](https://github.com/dotnet/machinelearning/pull/750#issuecomment-416349177) [](ancestors = 416349177,416348763)

**Comment by codemzs** [comment]
One PR for IDV that checks in baseline test(this). A second PR that migrates Parquet test from TLC repo to ML.NET.

---
In reply to: [416349633](https://github.com/dotnet/machinelearning/pull/750#issuecomment-416349633) [](ancestors = 416349633,416349177,416348763)

File: test/BaselineOutput/SingleDebug/Command/Datatypes-datatypes.txt [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+#@   col=i8:I8:4
+#@   col=ts:TS:5
+#@   col=dto:DZ:6
+#@   col=dt:DT:7
+#@   col=tx:TX:8
+#@ }
+bl	i1	i2	i4	i8	ts	dto	dt	tx
+0	127	32767	2147483647	9223372036854775807	"2.00:00:00"	"2008-11-30T00:00:00.0000000+00:00"	"2013-08-05T00:00:00.0000000"	foo
+1	-127	-32767	-2147483647	-9223372036854775807	"7.00:00:00"	"2008-11-30T00:00:00.0000000+00:00"	"2013-08-05T00:00:00.0000000"	xyz
+					"7.00:00:00"	"2008-11-30T00:00:00.0000000+00:00"	"2013-08-05T00:00:00.0000000"	xyz
```
Comment (TomFinley) [inline test/BaselineOutput/SingleDebug/Command/Datatypes-datatypes.txt]: >xyz [](start = 84, length = 3)

Maybe this one ought to be empty for text, since that is an interesting boundary case as well.

**Comment by TomFinley** [comment]
OK. That Parquet test will need to have integers in it for it to be meaningful.

---
In reply to: [416350157](https://github.com/dotnet/machinelearning/pull/750#issuecomment-416350157) [](ancestors = 416350157,416349633,416349177,416348763)

File: test/BaselineOutput/SingleDebug/Command/Datatypes-datatypes.txt (hunk 1 above)
Comment (codemzs) [reply test/BaselineOutput/SingleDebug/Command/Datatypes-datatypes.txt]: @tfinley@gmail.com The line below contains empty string. 

---
In reply to: [213097496](https://github.com/dotnet/machinelearning/pull/750#discussion_r213097496) [](ancestors = 213097496)

Comment (TomFinley) [reply test/BaselineOutput/SingleDebug/Command/Datatypes-datatypes.txt]: No, it does not. That is a *missing* string. Under default settings of the text loader and saver, the text field that has no content is considered missing, but with two quotes is considered the empty string. `""` vs. nothing.

File: test/Microsoft.ML.TestFramework/TestCommandBase.cs [hunk 2]
```diff
@@ -2028,5 +2028,14 @@ public void CommandTrainingBinaryFieldAwareFactorizationMachineWithValidationAnd
...
         }
+
+        [Fact]
+        public void DataTypes()
+        {
+            string idvPath = GetDataPath("datatypes.idv");
+            OutputPath textOutputPath = CreateOutputPath("datatypes.txt");
+            TestCore("savedata", idvPath, "loader=binary", "saver=text", textOutputPath.Arg("dout"));
+            Assert.True(textOutputPath.CheckEquality());
+        }
```
[the remark this answers is not in the document]
Comment (TomFinley) [reply test/Microsoft.ML.TestFramework/TestCommandBase.cs]: Right. We should firstly *not* be doing an `Assert.True` on check equality. The failure, if it happens, ought to be returned by the call to "Done" to allow a more comprehensive error. Also if there was a "soft failure" inside `TestCore`, we wouldn't be catching it *unless* we call `Done()`.

---
In reply to: [213145100](https://github.com/dotnet/machinelearning/pull/750#discussion_r213145100) [](ancestors = 213145100)

[the remark this answers is not in the document]
Comment (TomFinley) [reply test/Microsoft.ML.TestFramework/TestCommandBase.cs]: If we are intending this to be a baseline test, it would be more typical to make this a "Check". The trouble with `Assert.True` is that it will fail on the first failure, rather than collecting *all* baseline differences. This makes it harder to resolve failing tests. But for now maybe we keep the code as is, but yes this should have been improved.

Not sure how this happened. This was done deliberately, which means someone actually made the test hardness worse on purpose when we moved the code.

But you still need the call to `Done()`...

---
In reply to: [213148985](https://github.com/dotnet/machinelearning/pull/750#discussion_r213148985) [](ancestors = 213148985,213145100)

File: test/BaselineOutput/SingleDebug/Command/Datatypes-datatypes.txt [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+#@   col=ts:TS:5
+#@   col=dto:DZ:6
+#@   col=dt:DT:7
+#@   col=tx:TX:8
+#@ }
+bl	i1	i2	i4	i8	ts	dto	dt	tx
+0	127	32767	2147483647	9223372036854775807	"2.00:00:00"	"2008-11-30T00:00:00.0000000+00:00"	"2013-08-05T00:00:00.0000000"	foo
+1	-127	-32767	-2147483647	-9223372036854775807	"7.00:00:00"	"2008-11-30T00:00:00.0000000+00:00"	"2013-08-05T00:00:00.0000000"	xyz
+					"7.00:00:00"	"2008-11-30T00:00:00.0000000+00:00"	"2013-08-05T00:00:00.0000000"	
+9	8:
```
Comment (TomFinley) [inline test/BaselineOutput/SingleDebug/Command/Datatypes-datatypes.txt]: >8: [](start = 2, length = 2)

This indicates a missing value BTW.

File: test/Microsoft.ML.TestFramework/TestCommandBase.cs [hunk 4]
```diff
@@ -2028,5 +2028,15 @@ public void CommandTrainingBinaryFieldAwareFactorizationMachineWithValidationAnd
...
             Done();
         }
+
+        [Fact]
+        public void DataTypes()
+        {
+            string idvPath = GetDataPath("datatypes.idv");
+            OutputPath textOutputPath = CreateOutputPath("datatypes.txt");
+            TestCore("savedata", idvPath, "loader=binary", "saver=text", textOutputPath.Arg("dout"));
+            Contracts.Check(textOutputPath.CheckEquality());
```
Comment (TomFinley) [inline test/Microsoft.ML.TestFramework/TestCommandBase.cs]: >Contracts.Check [](start = 12, length = 15)

I think you just meant `Check`, not `Contracts.Check`. Also you'll want to include a sensible error message.

Comment (TomFinley) [reply test/Microsoft.ML.TestFramework/TestCommandBase.cs]: Also I think `TestCore` should be running the comparison itself (is it not) so I don't think this line serves any purpose.

---
In reply to: [213156468](https://github.com/dotnet/machinelearning/pull/750#discussion_r213156468) [](ancestors = 213156468)

File: test/Microsoft.ML.TestFramework/TestCommandBase.cs [hunk 5]
```diff
@@ -2028,5 +2029,18 @@ public void CommandTrainingBinaryFieldAwareFactorizationMachineWithValidationAnd
             Assert.True(outputPath.CheckEqualityNormalized());
             Done();
         }
+
+        [Fact]
+        public void DataTypes()
+        {
+            //Skip for linux because DATE/TIME format is different.
```
Comment (eerhardt) [inline test/Microsoft.ML.TestFramework/TestCommandBase.cs]: Can we instead write the DATE/TIME using a stable format that works across all platforms?
